File: Binary_Search_Tree/173.Binary_Search_Tree_Iterator/Python/173.Binary_Search_Tree_Iterator.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class BSTIterator(object):
    def __init__(self, root):
        """
        :type root: Optional[TreeNode]
        :rtype: None, void
        """
        ############
        #Initialize
        ##### Record array #####
        self.record_arry = []

        ##### Inorder traversal function call #####
        self.inorderTraversal(root)

        ##### Total length #####
        self.total_length = len(self.record_arry)

        ##### Record traversal pointer #####
        self.record_trav_ptr = 0
        

    def next(self):
        """
        :type: None, void
        :rtype: int
        """
        if (self.record_trav_ptr >= self.total_length): #Issue/Boundary-cae handling
            return (-1)

        ####################
        #Whole process/flow
        res_val = (self.record_arry)[self.record_trav_ptr] #Result value

        self.record_trav_ptr += 1 #Update/Traverse

        return res_val
        

    def hasNext(self):
        """
        :type: None, void
        :rtype: bool
        """
        if (self.record_trav_ptr >= self.total_length): #Issue/Boundary-cae handling
            return False

        return True

    
    def inorderTraversal(self, root):
        """
        :type root: Optional[TreeNode]
        :rtype: None, void
        """
        ############
        #Initialize
        ##### Record stack #####
        record_stack = []

        ####################
        #Whole process/flow
        while (root or record_stack):
            while (root):
                record_stack.append(root) #Keep updating/recording

                root = root.left #Keep updating/overwriting

            left_most_node = record_stack.pop() #Keep updating/overwriting

            (self.record_arry).append(left_most_node.val) #Keep updating/recording

            root = left_most_node.right #Keep updating/overwriting
        


# Your BSTIterator object will be instantiated and called as such:
# obj = BSTIterator(root)
# param_1 = obj.next()
# param_2 = obj.hasNext()

File: Binary_Search_Tree/173.Binary_Search_Tree_Iterator/C++/173.Binary_Search_Tree_Iterator.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class BSTIterator 
{
    /*Initialize*/
    ///// Total length /////
    int total_length;

    ///// Record traversal pointer /////
    int record_trav_ptr;

    ///// Record vector /////
    vector<int> record_vect;


public:
    BSTIterator(TreeNode * root) 
    {
        //=================================//
        // Input type:                     //
        //   - TreeNode * (root tree-node) //
        // Return type:                    //
        //   - void (no return)            //
        //=================================//

        /*Whole process, flow*/
        (this->inorderTraversal)(root, (this->record_vect)); //Iterative function call

        (this->total_length) = (this->record_vect).size(); //Update, Overwrite
        (this->record_trav_ptr) = 0; //Update, Overwrite
    }
    

    int next() 
    {
        //======================//
        // Input type:          //
        //   - void (no input)  //
        // Return type:         //
        //   - void (no return) //
        //======================//

        /*Whole process, flow*/
        if ((this->record_trav_ptr) >= total_length) { return (-1); } //Issue, Boundary-case handling

        return (this->record_vect)[((this->record_trav_ptr)++)];
    }
    

    bool hasNext() 
    {
        //======================//
        // Input type:          //
        //   - void (no input)  //
        // Return type:         //
        //   - void (no return) //
        //======================//

        /*Whole process, flow*/
        if ((this->record_trav_ptr) >= total_length) { return false; } //Issue, Boundary-case handling

        return true;
    }


    void inorderTraversal(TreeNode * root, vector<int> & record_vect)
    {
        //======================================//
        // Input type:                          //
        //   - TreeNode * (root tree-node)      //
        //   - vector<int> ref. (record vector) //
        // Return type:                         //
        //   - void (no return)                 //
        //======================================//

        /*Initialize*/
        ///// Record stack /////
        stack<TreeNode * > record_stack;

        /*Whole process, flow*/
        while ((root != nullptr) || (! record_stack.empty())) //Whole
        {
            while (root != nullptr) //Part
            {
                record_stack.emplace(root); //Keep updating, recording

                root = (root->left); //Keep updating, overwriting

            } //Part

            TreeNode * left_most_node = record_stack.top(); record_stack.pop(); //Keep updating, popped

            record_vect.emplace_back((left_most_node->val)); //Keep updating, recording

            root = (left_most_node->right); //Keep updating, overwriting

        } //Whole
    }
};

/**
 * Your BSTIterator object will be instantiated and called as such:
 * BSTIterator* obj = new BSTIterator(root);
 * int param_1 = obj->next();
 * bool param_2 = obj->hasNext();
 */
